### ISSUE TITLE
Script interface object constructors are not thread-safe

### ISSUE THREAD

**Issue by jngrad** [espressomd/espresso#4449]
**tl;dr:** Script interface objects on ranks >= 1 may be stale pointers or pointers to uninitialized objects. The bug typically manifests itself with a segmentation fault, or with a hanging MPI process (e.g. if the constructor runs a MPI reduction after a statement that can throw).

## Background and problem statement

Originally, ESPResSo was designed with global variables in mind. When instantiating a new algorithm object, for example the P3M electrostatics actor, the core constructor was only called on rank 0. The constructor would sanitize parameters and if necessary throw a C++ exception that would be converted to a Python exception. When no throw occurred, the global variable would be updated with the new parameters, and a broadcast would take place to synchronize the object on rank 0 with the same object on all other ranks.

When the `ScriptInterface` was created, a different pattern was used: a script interface `ObjerHandle` creation would be managed by a `ContextManager` that would rely on a `GlobalContext` resp. `LocalContext` to run the constructor on rank 0 resp. all other ranks. The `LocalContext` doesn't know about MPI communication and is registered as a callback by the `GlobalContext`. When the constructor throws, it has to throw an exception derived from `std::exception` on rank 0 and from `ScriptInterface::Exception` on the other ranks. Although the latter exception is silently ignored on the other ranks, it prevents the uninitialized object from being added to the `m_local_objects` registry:

https://github.com/espressomd/espresso/blob/7e29f9052e710fe1ebf0f5d2a8076b32921fbc6a/src/script_interface/GlobalContext.cpp#L46-L55

This design decision forces developers to guarantee that when the constructor throws an exception on rank 0, an exception of a different type must be thrown at the same point on all other ranks. Otherwise, the object is not created on rank 0 but is created on all the other ranks and stored in the `std::unordered_map` registry using the memory address of the object from rank 0 as key. Since the object is not created on rank 0, its dynamic memory is freed and the operating system is allowed to re-use that memory address if another memory allocation is requested with the same number of bytes, which typically happens when the user creates a new actor with corrected parameters. In that case a new object will be created on rank 0 and on the other ranks, but on the other ranks the new object will not be saved in the registry because it already contains another object with the same key (see #4448). Since that other object is either a stale reference, an uninitialized object or an initialized object with different parameters, undefined behavior is invoked.

Below is a printf trace obtained by instantiating a `DipolarP3M` object with incorrect parameters (to trigger an exception on rank 0 and store a stale reference on rank 1), and instantiating a fresh `DipolarP3M` object and passing it to the constructor of a `DLC` actor. Note how the `DLC` actor on rank 1 is passed the stale `DipolarP3M` sip (0x288bbe0) instead of the one that was just created (0x2853e90).
```
# python: instantiate a DipolarP3M object with wrong arguments to raise an error
# (::DipolarP3M core constructor throws an exception on all ranks, it's caught
# in the ScriptInterface::Dipole::DipolarP3M class and re-thrown on rank 0)
GlobalContext: rank 0: sip=0x1f8bac0 (constructing a ScriptInterface::Dipole::DipolarP3M)
LocalContext:  rank 1: sip=0x288bbe0 (constructing a ScriptInterface::Dipole::DipolarP3M)
# python: instantiate a valid DipolarP3M object and instantiate a DLC object with it
GlobalContext: rank 0: sip=0x1f8bac0 (constructing a ScriptInterface::Dipole::DipolarP3M)
LocalContext:  rank 1: sip=0x2853e90 (constructing a ScriptInterface::Dipole::DipolarP3M)
ScriptInterface::Dipole::DipolarP3M rank 1: core object successfully constructed; sip=0x2853e90 core=0x284ed30
ScriptInterface::Dipole::DipolarP3M rank 0: core object successfully constructed; sip=0x1f8bac0 core=0x1fb15c0
GlobalContext: rank 0: sip=0x1f52ac0 (constructing a ScriptInterface::Dipole::DLC)
LocalContext:  rank 1: sip=0x27fe4c0 (constructing a ScriptInterface::Dipole::DLC)
ScriptInterface::Dipole::DLC rank 0: input arguments:
  m_actor_dp3m.get()=0x1f8bac0
  m_actor_dp3m->core_actor()=0x1fb15c0
  m_actor_dp3m->m_tune_timings=10
ScriptInterface::Dipole::DLC rank 1: input arguments:
  m_actor_dp3m.get()=0x288bbe0
  m_actor_dp3m->core_actor()=nullptr
  m_actor_dp3m->m_tune_timings=42475136
[lama:921980] *** Process received signal ***
[lama:921980] Signal: Segmentation fault (11)
[lama:921980] Signal code: Address not mapped (1)
[lama:921980] Failing at address: 0x890
```

Replacing the `std::unordered_map<Key,T>::emplace()` by `std::unordered_map<Key,T>::operator[]` is not sufficient. For example a constructor may not fail on rank 0 but fail on rank 1, in which case rank 1 contains an uninitialized object. We need to be able to communicate that failure from rank 1 to rank 0.

## Proposed solution

Make the `LocalContext` MPI-aware, so that worker nodes can communication exceptions back to rank 0. In pseudo-code:
```c++
parallel try {
  m_obj = std::make_shared<CoreClass>(get_value<double>(params, "arg"));
} parallel catch (std::exception const &error) {
  handle_impl(error);
} parallel else {
  handle_impl();
}
```

This parallel try-catch-else uses two reductions to determine whether an exception occurred on rank 0, and if not, whether an exception occurred on any other rank. When an exception is thrown in the `CoreClass` constructor, two outcomes are possible:
* rank 0 failed: re-throw exception on rank 0, throw `ScriptInterface::Exception` on other ranks
* rank 0 didn't fail, but at least one other rank failed: collect the output of `std::exception::what()` from all ranks and gather on rank 0 where a `std::runtime_error` with a summary of the exceptions is thrown, other ranks throw `ScriptInterface::Exception`

The role of `ScriptInterface::Exception` is to guarantee that no object is ever added to the registry when construction failed on any rank.

**Comment by jngrad** [comment]
The proposed parallel script interface handler is now used in the collision detection, cell system, electrostatics and magnetostatics interfaces. It will be rolled out to the other features as we gradually replace the `MpiCallbacks` framework by regular `boost::mpi` communication.
